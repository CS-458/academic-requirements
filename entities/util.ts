let id = 0;
export function uniqueId(): string {
  return `id_${id++}`;
}
