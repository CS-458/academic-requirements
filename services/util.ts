/// Fetch a request (similar to fetch), but throw an error if the request fails
export async function fetchApi<T>(url: string, init?: RequestInit): Promise<T> {
  const res = await fetch(url, init);
  if (!res.ok) {
    throw new Error(`Response code ${res.status} for request to ${url}`);
  }
  return await res.json();
}
