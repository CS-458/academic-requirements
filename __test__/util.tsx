import { jest } from "@jest/globals";
import type { MatcherFunction } from "expect";
import { DndProvider } from "react-dnd";
import { HTML5Backend } from "react-dnd-html5-backend";
import { QueryClient, QueryClientProvider, UseQueryResult } from "react-query";
import { testApiHandler } from "next-test-api-route-handler";
import { NextApiHandler } from "next";
import {
  Matcher,
  screen,
  render as testRender,
  RenderResult,
  waitFor
} from "@testing-library/react";
import userEvent from "@testing-library/user-event";
import { UserEvent } from "@testing-library/user-event/dist/types/setup/setup";
import { User, UserMajor } from "../services/user";
import { VerifyIdTokenOptions, LoginTicket } from "google-auth-library";
import { updateClient } from "../services/login";
import { PromisedDatabase } from "promised-sqlite3";
import { setUserDb } from "../services/sql";

import "../pages/api";
import { dragAndDrop } from "./dragDrop";

interface UserExt {
  /// Select an element from an `AutoComplete` dropdown
  selectAutocomplete: typeof selectAutocomplete;
  /// Drag an element from one place to another
  drag: typeof dragAndDrop;
}

async function selectAutocomplete(
  this: UserEvent,
  label: Matcher,
  option: Matcher
): Promise<void> {
  await this.click(screen.getByTestId("test-root-element"));
  const input = screen.getByLabelText(label, { selector: "input" });
  await waitFor(async () => {
    expect(input).not.toBeDisabled();
  });
  await this.click(input);
  try {
    await this.click(
      await screen.findByText(option, undefined, { timeout: 10000 })
    );
  } catch (e) {
    const rootEl = screen.getByTestId("test-root-element");
    const popper = rootEl?.querySelector(".MuiAutocomplete-popper");
    const html = popper?.innerHTML;
    throw new Error(
      `Option ${option} was not found in dropdown.\nOptions: ${html}`
    );
  }
}

export function parentEl(el: HTMLElement, name: string): HTMLElement {
  while (!el.classList.contains(`${name}-root`)) {
    const next = el.parentElement;
    if (next == null) {
      throw new Error(`Element does not have a wrapper with name ${name}`);
    }
    el = next;
  }
  return el;
}

/// userEvent.setup(), but with some additional convience methods
export function setupUser(): UserEvent & UserExt {
  return {
    ...userEvent.setup(),
    selectAutocomplete,
    drag: dragAndDrop
  };
}

// const modules: { [key: string]: NextApiHandler | Promise<NextApiHandler> } = {};
// declare function fetch2(input: RequestInfo | URL, init?: RequestInit): Promise<Response>;
/// Executes a request against an Api Route, roughly equavelent with `fetch`.
export async function fetchApiRoute(
  url: RequestInfo | URL,
  init?: RequestInit
): Promise<Response> {
  const res = { resolve: (_a: any) => { }, reject: (_e: any) => { } };
  const result: Promise<Response> = new Promise((resolve, reject) => {
    res.resolve = resolve;
    res.reject = reject;
  });

  let fullUrl: string | undefined;
  let path;
  let params;
  if (url instanceof URL) {
    fullUrl = url.href;
    path = url.pathname;
    params = url.search;
  } else if (typeof url === "string") {
    fullUrl = url;
    [path, params] = url.split("?");
  } else {
    fullUrl = url.url;
    [path, params] = url.url.split("?");
  }

  const handler: NextApiHandler = await import(`../pages/${path}`).catch(
    (e) => {
      if (jest.isEnvironmentTornDown()) {
        console.log("Request made after environment torn down");
      } else {
        console.error(e);
      }
    }
  );
  await testApiHandler({
    url: fullUrl,
    handler,
    params: params?.split("&").reduce((a, b) => {
      const [name, value] = b.split("=");
      return { ...a, [name]: value };
    }, {}),
    test: async ({ fetch }) => {
      const inner = await fetch(init);
      res.resolve(inner);
    }
  });
  return await result;
}

/// Calls fetchApiRoute, and decodes the response as JSON
export async function fetchApiJson(
  url: RequestInfo | URL,
  init?: RequestInit
): Promise<any> {
  return await (await fetchApiRoute(url, init)).json();
}

/**
 *  Wrap Elements with QueryClient & DndProvider similar to _app.tsx
 *
 *  Also overides window.fetch to use fetchApiRoute for testing
 * */
export function render(children: JSX.Element | JSX.Element[]): RenderResult {
  window.fetch = fetchApiRoute;
  return testRender(
    <div data-testid="test-root-element">
      <QueryClientProvider client={new QueryClient()}>
        <DndProvider backend={HTML5Backend}>{children}</DndProvider>
      </QueryClientProvider>
    </div>
  );
}

const localStorage: { [key: string]: string | null } = {};

export function getMockStorage(key: string): string | null {
  return localStorage[key];
}

export function setMockStorage(key: string, value: string): void {
  localStorage[key] = value;
}

export function removeMockStorage(key: string): void {
  localStorage[key] = null;
}

export function clearMockStorage(): void {
  for (const k in localStorage) {
    localStorage[k] = null;
  }
}

/// Configure local storage and set pre-existing data in local storage
export function buildLocalStorage(user?: UserMajor): void {
  window.localStorage.getItem = getMockStorage;
  window.localStorage.setItem = setMockStorage;
  window.localStorage.removeItem = removeMockStorage;
  window.localStorage.clear = clearMockStorage;
  if (user !== undefined) {
    window.localStorage.setItem("user_major", JSON.stringify(user));
  }
}

export function mockUseQuery<T>(
  _key: any,
  fetch: () => Promise<T>
): UseQueryResult<T> & { fetch: Promise<T> } {
  return {
    fetch: fetch(),
    data: undefined,
    error: null,
    isError: false,
    isIdle: false,
    isLoading: true,
    isLoadingError: false,
    isRefetchError: false,
    isSuccess: false,
    status: "loading",
    dataUpdatedAt: Date.now(),
    errorUpdatedAt: Date.now(),
    failureCount: 0,
    errorUpdateCount: 0,
    isFetched: true,
    isFetchedAfterMount: true,
    isFetching: false,
    isPlaceholderData: false,
    isPreviousData: false,
    isRefetching: false,
    isStale: false,
    refetch: async (_options: any) => {
      throw new Error("");
    },
    remove: () => { }
  };
}

export function setupStandaloneQuery(): void {
  jest.mock("react-query", () => ({
    useQuery: mockUseQuery
  }));
}

export async function waitForQuery<T>(
  query: () => UseQueryResult<T>
): Promise<T> {
  // @ts-expect-error fetch only exists on our mock
  const temp = await query().fetch;
  return temp;
}

const toBeUnique: MatcherFunction<[]> = (actual) => {
  if (!Array.isArray(actual)) {
    throw new Error("actual is not an object");
  }
  if (actual === null) {
    return {
      message: () => "null is not an array",
      pass: false
    };
  }
  for (const i of actual) {
    const pos = actual.find((val, idx) => idx > i && actual[i] === val);
    if (pos !== undefined) {
      return {
        message: () => `Indicies ${i}, and ${pos} are the same`,
        pass: false
      };
    }
  }
  return {
    message: () => "All elements are unique",
    pass: true
  };
};

const toBeHidden: MatcherFunction<[]> = (actual) => {
  if (!(actual instanceof HTMLElement)) {
    throw new Error("actual is not an HTMLElement");
  }
  let cur = actual;
  console.log("child", cur.ariaHidden);
  while (cur.ariaHidden == null || cur.ariaHidden === "false") {
    if (cur.dataset.testid === "Semester") {
      console.log(cur.ariaHidden);
    }
    const next = cur.parentElement;
    if (next === null) {
      return {
        message: () => "Element (or parent) is not aria-hidden",
        pass: false
      };
    }
    cur = next;
  }
  return {
    message: () => "Element (or parent) is aria-hidden",
    pass: true
  };
};

expect.extend({
  toBeUnique,
  toBeHidden
});

declare module "expect" {
  interface AsymmetricMatchers {
    toBeUnique: () => void;
    toBeHidden: () => void;
  }
  interface Matchers<R> {
    toBeUnique: () => R;
    toBeHidden: () => R;
  }
}

export function courseSemestersCheck(semesters: string): void {
  /*
    Available options: FA,WI,SP,SU

    semesters can start with any of the options
    then be followed by a comma and any of the options

    Note: Data such as FA,FA will match but FA,bad_data will not
  */
  // Expect the format to match
  expect(semesters).toMatch(/^(FA|WI|SP|SU)(,(FA|WI|SP|SU))*$/);
}

export function createMockToken(): void {
  updateClient((c) => {
    c.verifyIdToken = async (options: VerifyIdTokenOptions) => {
      const [test, id] = options.idToken.split(":");
      if (test !== "TEST_TOKEN") {
        throw new Error("Invalid Token");
      }
      return new LoginTicket("", {
        sub: id,
        aud: c._clientId ?? "",
        exp: Date.now() + 1000000,
        iat: Date.now() - 100,
        iss: "https://accounts.google.com",
        at_hash: ""
      });
    };
  });
}

export function mockToken(id: string): string {
  return `TEST_TOKEN:${id}`;
}

export function mockUserInfo(id: string): User {
  return {
    cred: mockToken(id),
    info: {
      sub: id,
      picture: "",
      name: id,
      email: `${id}@gmail.com`
    }
  };
}

const db = {
  db: new PromisedDatabase(),
  opened: false
};

export async function setupMockUserDB(): Promise<void> {
  createMockToken();
  setUserDb(db.db);
  if (!db.opened) {
    db.opened = true;
    await db.db.open(":memory:");
    await db.db.run(
      "CREATE TABLE `user` ( `idUser` TEXT PRIMARY KEY, `role` TEXT NOT NULL)"
    );
    await db.db.run(
      "CREATE TABLE `schedule` ( `userID` TEXT, `name` TEXT, `timestamp` DATE DEFAULT (datetime('now','localtime')), `scheduleData` TEXT NOT NULL, PRIMARY KEY(`userID`, `name`))"
    );
  }
}
