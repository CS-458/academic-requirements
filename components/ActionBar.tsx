/*
  Nick Raffel
  This is just a component for the Action Bar on the left Side that houses the save button
*/

import React from "react";
import { Box } from "@mui/material";
import ScheduleUpload from "./ScheduleUploadModal";
import { UserSavedSchedule } from "../entities/four_year_plan";

// Schedule Data and SetAlertData are being passed through here into the Schedule Upload Modal
export default function BoxSx(props: {
  scheduleData: UserSavedSchedule["scheduleData"],
  setAlertData: (msg: string, severity: string) => void
}): any {
  return (
    <div>
      <Box
        sx={{
          height: "100%",
          backgroundColor: "white",
          paddingTop: "2em",
          boxShadow: 9
        }}
      >
        <ScheduleUpload
          scheduleData={props.scheduleData}
          setAlertData={props.setAlertData}
        />
      </Box>
    </div>
  );
}
