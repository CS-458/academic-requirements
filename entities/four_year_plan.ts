import type { TargetType } from "dnd-core";
import { CourseError } from "../components/FourYearPlanPage";

// Defines the major object we are passing between files
export interface MajorType {
  id: number; // database major id
  name: string; // name of the major
}

// Defines the concentration object we are passing between files
export interface ConcentrationType {
  idConcentration: number; // Database concentration id
  name: string; // concentration name
  fourYearPlan: string; // this is a json of a fourYearPlan stored in a string
}

// Defines the course object as it is used in most locations
export interface CourseType {
  credits: number; // number of credits a course has
  name: string; // course name
  number: string; // course number STORED IN DATABASE AS STRING
  semesters: string; // whether it runs in fall or spring (FA or SP) or null (runs both)
  subject: string; // course subject i.e. CS, MATH
  preReq: string; // string containing prerequisite courses in a special format
  category: string; // the category the course falls into
  idCourse: number; // Database course id
  idCategory: number; // Database id for the category
  dragSource: string; // Which element it came from i.e. "Semester 1" or CourseList
  repeatableForCred: boolean; // If a course can be taken twice for credit
}

// Defines the actual draggable course object type
export interface DragCourseType {
  name: string; // see CourseType
  subject: string; // see CourseType
  number: string; // see CourseType
  type: string; // Don't worry about this, all assigned to course
  credits: number; // see CourseType
  semesters: string; // see CourseType
  preReq: string; // see CourseType
  dragSource: string; // see CourseType
  warningYellowColor: number | undefined; // If the course is in multiple semesters
  warningOrangeColor: number | undefined; // If the course is in the wrong semester fall vs spring
  warningRedColor: number | undefined; // If the course has a prereq error
  idCourse: number; // see CourseType
  idCategory: number; // see CourseType
  repeatableForCred: boolean; // see CourseType
}

// Defines the props for the course list component
export interface CourseListType {
  accept: TargetType; // makes it a drop target
  onDrop: (item: any) => void; // function for what to do on drop
  courses: CourseType[]; // list of courses it displays for dragging
}

// Defines a requirement as it comes from the database
export interface RequirementType {
  courseCount: number; // number of courses required to satisfy
  courseReqs: string; // list of courses required to satisfy
  creditCount: number; // number of credits required to satisfy
  idCategory: number; // Database id for requirement
  name: string; // name of the requirement
  parentCategory: number; // Database id for parent requirement category
}

// Defines the saved information for each requirement
export interface RequirementComponentType {
  courseCount: number | null; // number of courses required to satisfy
  courseReqs: string | null; // list of courses required to satisfy
  creditCount: number; // number of credits required to satisfy
  idCategory: number; // Database id for requirement
  name: string; // name of the requirement
  shortName: string; // abbreviated version of the name for display
  parentCategory: number | null; // Database id for parent requirement category
  percentage: number; // Percentage that requirement has been filled
  inheritedCredits: number; // Number of credits a parent req inherits from child req
  coursesTaken: string; // list of courses already used toward requirement
  courseCountTaken: number; // number of courses taken towards requirement
  creditCountTaken: number; // number of credits taken towards requirement
}

// Defines a list of Requirements
export interface RequirementsType {
  [category: number]: RequirementType;
}

// Defines a list of courses in more than one category for requirements
export interface MultipleCategoriesType {
  categories: number[]; // list of category IDs the course falls under
  idString: string; // an identifier for the course i.e. CS-144
}

// define what can be a season
export enum season {
  Fall = "Fall",
  Winter = "Winter",
  Spring = "Spring",
  Summer = "Summer"
}
// define
export enum warning {
  Low = "Low",
  High = "High"
}

export function seasonNum(s: season): number {
  switch (s) {
    case season.Fall:
      return 0;
    case season.Winter:
      return 1;
    case season.Spring:
      return 2;
    case season.Summer:
      return 3;
  }
}

// Defines the properties of a semester component
export interface SemesterProps {
  accept: TargetType; // makes it a drop target, don't worry about it
  onDrop: (item: any) => void; // function for when an item is dropped
  semesterNumber: number; // the number of the semester
  courses: CourseType[]; // a list of courses in the semester
  SemesterCredits: number; // total credits of all courses in semester
  Warning: warning | null; // credit warning (low or high)
  warningPrerequisiteCourses: CourseError[]; // list of courses in with prereq issues
  warningFallvsSpringCourses: CourseError[];
  warningDuplicateCourses: CourseError[]; // list of courses in more than one semester
  year: number; // year number 1,2,3,4,etc.
  season: season; // Season the semester is (Fall, Winter, Spring, Summer)
}

// Defines the properties of a semester object
export interface SemesterType {
  accepts: string[]; // this is just a course constant
  semesterNumber: number; // number of the semester
  courses: CourseType[]; // list of courses in semester
  SemesterCredits: number; // number of credits in the semester
  Warning: warning | null; // credit warning (high or low)
  year: number; // year number 1,2,3,4,etc.
  season: season; // Season the semester is (Fall, Winter, Spring, Summer)
}

export function sortSemester(a: SemesterType, b: SemesterType): number {
  const aNum = seasonNum(a.season) + a.year * 4;
  const bNum = seasonNum(b.season) + b.year * 4;
  return aNum - bNum;
}

//  Defines the properties that should be passed in to the fourYearPlan page
export interface FourYearPlanType {
  PassedCourseList: CourseType[];
  requirements: RequirementComponentType[] | null | undefined;
  requirementsGen: RequirementComponentType[] | null | undefined;
  importData?: {};
}

export interface UserSavedSchedule {
  userID: string;
  name: string;
  timestamp: string;
  scheduleData: {
    Major: number;
    Concentration: number;
    "Completed Courses": string[];
    schedule: ScheduleData;
  };
}

export interface movedCourse {
  movedTo: number,
  movedFrom: number,
  course: number
}
export type ScheduleData = Array<{
  year: number;
  seasons: Array<{ season: season; classes: string[] }>;
}>;
