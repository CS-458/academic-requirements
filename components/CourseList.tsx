import type { CSSProperties, FC } from "react";
import React, { memo } from "react";
import { useDrop } from "react-dnd";
import { ItemTypes } from "../entities/Constants";
import { Course } from "./DraggableCourse";

// Styling for the course list
const style: CSSProperties = {
  height: "30rem",
  width: "90%",
  marginRight: ".5rem",
  marginBottom: ".5rem",
  color: "white",
  padding: "1rem",
  textAlign: "center",
  fontSize: "1rem",
  lineHeight: "normal",
  float: "left",
  background: "#004990",
  borderRadius: ".5rem",
  overflow: "auto"
};

export interface CourseListProps {
  accept: Course;
  onDrop: (item: any) => void;
  courses: Course[];
}

export const CourseList: FC<CourseListProps> = memo(function CourseList({
  accept,
  onDrop,
  courses
}) {
  // defines the drop
  const [{ isOver }, drop] = useDrop({
    accept,
    drop: onDrop,
    collect: (monitor) => ({
      isOver: monitor.isOver()
    })
  });

  // changes the background color on hover over course list
  const isActive = isOver;
  let backgroundColor = "#004990";
  if (isActive) {
    backgroundColor = "darkgreen";
  }

  return (
    <div ref={drop} style={{ ...style, backgroundColor }}>
      {isActive ? "Release to drop" : "CourseList"}
      {courses.map(
        ({ name, subject, number, semesters, credits, preReq }, index) => (
          <Course
            name={name}
            subject={subject}
            number={number}
            semesters={semesters}
            credits={credits}
            type={ItemTypes.COURSE}
            key={index}
            dragSource={"CourseList"}
            preReq={preReq}
          />
        )
      )}
    </div>
  );
});
