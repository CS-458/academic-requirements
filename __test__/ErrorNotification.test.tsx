import "@testing-library/jest-dom";
import { fireEvent, screen, within } from "@testing-library/react";
import { CourseType } from "../entities/four_year_plan";
import { setupUser, render, parentEl } from "./util";
import { FourYearPlanPage } from "../components/FourYearPlanPage";

const mockCourses: CourseType[] = [
  {
    subject: "SUB",
    number: "100",
    name: "Course 1",
    semesters: "FA",
    credits: 1,
    preReq: "",
    idCourse: 1,
    repeatableForCred: false,
    category: "CAT-1",
    idCategory: 1,
    dragSource: "CourseList"
  },
  {
    subject: "SUB",
    number: "200",
    name: "Course 2",
    semesters: "",
    credits: 2,
    preReq: "SUB_100",
    idCourse: 2,
    repeatableForCred: false,
    category: "CAT-1",
    idCategory: 2,
    dragSource: "CourseList"
  }
];
test("Four Year Plan Page renders", async () => {
  const index = render(
    <FourYearPlanPage
      PassedCourseList={mockCourses}
      requirements={undefined}
      requirementsGen={undefined}
      importData={undefined}
    />
  );
  expect(index.baseElement).toMatchSnapshot();
});

test("Four Year Plan Page duplicate courses", async () => {
  const user = setupUser();
  const index = render(
    <FourYearPlanPage
      PassedCourseList={mockCourses}
      requirements={undefined}
      requirementsGen={undefined}
      importData={undefined}
    />
  );
  expect(index.baseElement).toMatchSnapshot();
  const categoryTab = screen.getByTestId("category-tab");
  await user.click(categoryTab);
  await user.selectAutocomplete("Course Category", mockCourses[0].category);

  const firstYear = parentEl(screen.getByText(/Year 1/i), "MuiAccordion");
  expect(firstYear).toBeInTheDocument();
  const fallSemester = parentEl(
    within(firstYear).getByText(/Fall/i),
    "Semester"
  );
  expect(fallSemester).toBeInTheDocument();
  const springSemester = parentEl(
    within(firstYear).getByText(/Spring/i),
    "Semester"
  );
  expect(fallSemester).toBeInTheDocument();

  const course = screen.getAllByTestId("course");
  fireEvent.dragStart(course[0]);
  fireEvent.dragEnter(fallSemester);
  fireEvent.dragOver(fallSemester);
  fireEvent.drop(fallSemester);

  fireEvent.dragStart(course[0]);
  fireEvent.dragEnter(springSemester);
  fireEvent.dragOver(springSemester);
  fireEvent.drop(springSemester);

  expect(screen.getByTestId("snackbar")).toBeInTheDocument();
});

test("Four Year Plan Page Prerequisite course", async () => {
  const user = setupUser();
  const index = render(
    <FourYearPlanPage
      PassedCourseList={mockCourses}
      requirements={undefined}
      requirementsGen={undefined}
      importData={undefined}
    />
  );
  expect(index.baseElement).toMatchSnapshot();
  const categoryTab = screen.getByTestId("category-tab");
  await user.click(categoryTab);
  await user.selectAutocomplete("Course Category", mockCourses[0].category);

  const firstYear = parentEl(screen.getByText(/Year 1/i), "MuiAccordion");
  expect(firstYear).toBeInTheDocument();
  const fallSemester = parentEl(
    within(firstYear).getByText(/Fall/i),
    "Semester"
  );
  expect(fallSemester).toBeInTheDocument();

  const course = screen.getAllByTestId("course");
  fireEvent.dragStart(course[1]);
  fireEvent.dragEnter(fallSemester);
  fireEvent.dragOver(fallSemester);
  fireEvent.drop(fallSemester);
  expect(screen.getByTestId("snackbar")).toBeInTheDocument();
});

test("Four Year Plan incompatible Season", async () => {
  const user = setupUser();
  const index = render(
    <FourYearPlanPage
      PassedCourseList={mockCourses}
      requirements={undefined}
      requirementsGen={undefined}
      importData={undefined}
    />
  );
  expect(index.baseElement).toMatchSnapshot();
  const categoryTab = screen.getByTestId("category-tab");
  await user.click(categoryTab);
  await user.selectAutocomplete("Course Category", mockCourses[0].category);

  const firstYear = parentEl(screen.getByText(/Year 1/i), "MuiAccordion");
  expect(firstYear).toBeInTheDocument();
  const springSemester = parentEl(
    within(firstYear).getByText(/Spring/i),
    "Semester"
  );
  expect(springSemester).toBeInTheDocument();

  const course = screen.getAllByTestId("course");
  fireEvent.dragStart(course[0]);
  fireEvent.dragEnter(springSemester);
  fireEvent.dragOver(springSemester);
  fireEvent.drop(springSemester);
  expect(screen.getByTestId("snackbar")).toBeInTheDocument();
});
