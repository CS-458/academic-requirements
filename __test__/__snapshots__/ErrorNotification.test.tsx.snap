// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Four Year Plan Page renders 1`] = `
<body>
  <div>
    <div
      data-testid="test-root-element"
    >
      <div>
        <div
          class="drag-drop"
        >
          <div
            style="overflow: hidden; clear: both;"
          >
            <div
              data-testid="semester"
              style="height: 15rem; width: 19%; margin-right: .5rem; margin-bottom: .5rem; color: white; padding: 1rem; text-align: center; font-size: 1rem; line-height: normal; float: left; white-space: pre; background: rgb(0, 73, 144); border-radius: .5rem; overflow: auto; background-color: rgb(0, 73, 144);"
            >
              Semester 1 
 Fall 
 Credits 0
            </div>
            <div
              data-testid="semester"
              style="height: 15rem; width: 19%; margin-right: .5rem; margin-bottom: .5rem; color: white; padding: 1rem; text-align: center; font-size: 1rem; line-height: normal; float: left; white-space: pre; background: rgb(0, 73, 144); border-radius: .5rem; overflow: auto; background-color: rgb(0, 73, 144);"
            >
              Semester 2 
 Spring 
 Credits 0
            </div>
            <div
              data-testid="semester"
              style="height: 15rem; width: 19%; margin-right: .5rem; margin-bottom: .5rem; color: white; padding: 1rem; text-align: center; font-size: 1rem; line-height: normal; float: left; white-space: pre; background: rgb(0, 73, 144); border-radius: .5rem; overflow: auto; background-color: rgb(0, 73, 144);"
            >
              Semester 3 
 Fall 
 Credits 0
            </div>
            <div
              data-testid="semester"
              style="height: 15rem; width: 19%; margin-right: .5rem; margin-bottom: .5rem; color: white; padding: 1rem; text-align: center; font-size: 1rem; line-height: normal; float: left; white-space: pre; background: rgb(0, 73, 144); border-radius: .5rem; overflow: auto; background-color: rgb(0, 73, 144);"
            >
              Semester 4 
 Spring 
 Credits 0
            </div>
            <div
              data-testid="semester"
              style="height: 15rem; width: 19%; margin-right: .5rem; margin-bottom: .5rem; color: white; padding: 1rem; text-align: center; font-size: 1rem; line-height: normal; float: left; white-space: pre; background: rgb(0, 73, 144); border-radius: .5rem; overflow: auto; background-color: rgb(0, 73, 144);"
            >
              Semester 5 
 Fall 
 Credits 0
            </div>
            <div
              data-testid="semester"
              style="height: 15rem; width: 19%; margin-right: .5rem; margin-bottom: .5rem; color: white; padding: 1rem; text-align: center; font-size: 1rem; line-height: normal; float: left; white-space: pre; background: rgb(0, 73, 144); border-radius: .5rem; overflow: auto; background-color: rgb(0, 73, 144);"
            >
              Semester 6 
 Spring 
 Credits 0
            </div>
            <div
              data-testid="semester"
              style="height: 15rem; width: 19%; margin-right: .5rem; margin-bottom: .5rem; color: white; padding: 1rem; text-align: center; font-size: 1rem; line-height: normal; float: left; white-space: pre; background: rgb(0, 73, 144); border-radius: .5rem; overflow: auto; background-color: rgb(0, 73, 144);"
            >
              Semester 7 
 Fall 
 Credits 0
            </div>
            <div
              data-testid="semester"
              style="height: 15rem; width: 19%; margin-right: .5rem; margin-bottom: .5rem; color: white; padding: 1rem; text-align: center; font-size: 1rem; line-height: normal; float: left; white-space: pre; background: rgb(0, 73, 144); border-radius: .5rem; overflow: auto; background-color: rgb(0, 73, 144);"
            >
              Semester 8 
 Spring 
 Credits 0
            </div>
            <button
              data-testid="ExportButton"
            >
              Export Schedule
            </button>
          </div>
          <div
            class="class-dropdown"
            style="overflow: hidden; clear: both;"
          >
            <div>
              <div
                class="course-box-header"
              >
                <div
                  class="MuiAutocomplete-root MuiAutocomplete-hasPopupIcon css-1pskylb-MuiAutocomplete-root"
                >
                  <div
                    class="MuiFormControl-root MuiFormControl-fullWidth MuiTextField-root css-wb57ya-MuiFormControl-root-MuiTextField-root"
                  >
                    <div
                      class="MuiInputBase-root MuiOutlinedInput-root MuiInputBase-colorPrimary MuiInputBase-fullWidth MuiInputBase-formControl MuiInputBase-adornedEnd MuiAutocomplete-inputRoot css-154xyx0-MuiInputBase-root-MuiOutlinedInput-root"
                    >
                      <input
                        aria-autocomplete="list"
                        aria-expanded="false"
                        aria-invalid="false"
                        autocapitalize="none"
                        autocomplete="off"
                        class="MuiInputBase-input MuiOutlinedInput-input MuiInputBase-inputAdornedEnd MuiAutocomplete-input MuiAutocomplete-inputFocused css-nxo287-MuiInputBase-input-MuiOutlinedInput-input"
                        id=":r0:"
                        role="combobox"
                        spellcheck="false"
                        type="text"
                        value=""
                      />
                      <div
                        class="MuiAutocomplete-endAdornment css-1q60rmi-MuiAutocomplete-endAdornment"
                      >
                        <button
                          aria-label="Open"
                          class="MuiButtonBase-root MuiIconButton-root MuiIconButton-sizeMedium MuiAutocomplete-popupIndicator css-qzbt6i-MuiButtonBase-root-MuiIconButton-root-MuiAutocomplete-popupIndicator"
                          tabindex="-1"
                          title="Open"
                          type="button"
                        >
                          <svg
                            aria-hidden="true"
                            class="MuiSvgIcon-root MuiSvgIcon-fontSizeMedium css-i4bv87-MuiSvgIcon-root"
                            data-testid="ArrowDropDownIcon"
                            focusable="false"
                            viewBox="0 0 24 24"
                          >
                            <path
                              d="M7 10l5 5 5-5z"
                            />
                          </svg>
                          <span
                            class="MuiTouchRipple-root css-8je8zh-MuiTouchRipple-root"
                          />
                        </button>
                      </div>
                      <fieldset
                        aria-hidden="true"
                        class="MuiOutlinedInput-notchedOutline css-1d3z3hw-MuiOutlinedInput-notchedOutline"
                      >
                        <legend
                          class="css-ihdtdm"
                        >
                          <span
                            class="notranslate"
                          >
                            â€‹
                          </span>
                        </legend>
                      </fieldset>
                    </div>
                  </div>
                </div>
              </div>
            </div>
            <div
              style="height: 30rem; width: 90%; margin-right: .5rem; margin-bottom: .5rem; color: white; padding: 1rem; text-align: center; font-size: 1rem; line-height: normal; float: left; background: rgb(0, 73, 144); border-radius: .5rem; overflow: auto; background-color: rgb(0, 73, 144);"
            />
          </div>
          <div
            class="right-information-box"
          >
            <div
              class="right-information-box-header"
            >
              <div
                class="MuiAutocomplete-root MuiAutocomplete-hasPopupIcon css-1pskylb-MuiAutocomplete-root"
              >
                <div
                  class="MuiFormControl-root MuiFormControl-fullWidth MuiTextField-root css-wb57ya-MuiFormControl-root-MuiTextField-root"
                >
                  <div
                    class="MuiInputBase-root MuiOutlinedInput-root MuiInputBase-colorPrimary MuiInputBase-fullWidth MuiInputBase-formControl MuiInputBase-adornedEnd MuiAutocomplete-inputRoot css-154xyx0-MuiInputBase-root-MuiOutlinedInput-root"
                  >
                    <input
                      aria-autocomplete="list"
                      aria-expanded="false"
                      aria-invalid="false"
                      autocapitalize="none"
                      autocomplete="off"
                      class="MuiInputBase-input MuiOutlinedInput-input MuiInputBase-inputAdornedEnd MuiAutocomplete-input MuiAutocomplete-inputFocused css-nxo287-MuiInputBase-input-MuiOutlinedInput-input"
                      id=":r2:"
                      role="combobox"
                      spellcheck="false"
                      type="text"
                      value=""
                    />
                    <div
                      class="MuiAutocomplete-endAdornment css-1q60rmi-MuiAutocomplete-endAdornment"
                    >
                      <button
                        aria-label="Open"
                        class="MuiButtonBase-root MuiIconButton-root MuiIconButton-sizeMedium MuiAutocomplete-popupIndicator css-qzbt6i-MuiButtonBase-root-MuiIconButton-root-MuiAutocomplete-popupIndicator"
                        tabindex="-1"
                        title="Open"
                        type="button"
                      >
                        <svg
                          aria-hidden="true"
                          class="MuiSvgIcon-root MuiSvgIcon-fontSizeMedium css-i4bv87-MuiSvgIcon-root"
                          data-testid="ArrowDropDownIcon"
                          focusable="false"
                          viewBox="0 0 24 24"
                        >
                          <path
                            d="M7 10l5 5 5-5z"
                          />
                        </svg>
                        <span
                          class="MuiTouchRipple-root css-8je8zh-MuiTouchRipple-root"
                        />
                      </button>
                    </div>
                    <fieldset
                      aria-hidden="true"
                      class="MuiOutlinedInput-notchedOutline css-1d3z3hw-MuiOutlinedInput-notchedOutline"
                    >
                      <legend
                        class="css-ihdtdm"
                      >
                        <span
                          class="notranslate"
                        >
                          â€‹
                        </span>
                      </legend>
                    </fieldset>
                  </div>
                </div>
              </div>
            </div>
            <div
              class="right-information-box-content"
            >
              <p
                class="right-information-box-description"
              >
                Select a category and drag a course onto a semester to begin planning.
              </p>
            </div>
          </div>
        </div>
      </div>
    </div>
  </div>
</body>
`;
